qiskit/transpiler/passes/decoupling/insert_dd.py

```diff
@@ -0,0 +1,100 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Dynamical Decoupling insertion pass."""
+from collections import defaultdict
+from typing import List
+import numpy as np
+
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.circuit import Delay
+from qiskit.circuit.library import U3Gate, XGate
+from qiskit.transpiler.passes import Optimize1qGates
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+class InsertDD(TransformationPass):
+    """DD insertion pass."""
+
+    def __init__(self, durations, dd_sequence):
+        """ASAPSchedule initializer.
+        Args:
+            durations (InstructionDurations): Durations of instructions to be used in scheduling
+            dd_sequence (list[str]): sequence of gates to apply in idle spots
+        """
+        super().__init__()
+        self._durations = durations
+        self._dd_sequence = dd_sequence
+
+    def run(self, dag):
+        """Run the InsertDD pass on `dag`.
+        Args:
+            dag (DAGCircuit): a scheduled DAG.
+        Returns:
+            DAGCircuit: equivalent circuit with no extra delays but DD where possible.
+        Raises:
+            TranspilerError: if the circuit is not mapped on physical qubits.
+        """
+        if len(dag.qregs) != 1 or dag.qregs.get('q', None) is None:
+            raise TranspilerError('DD runs on physical circuits only')
+
+        new_dag = DAGCircuit()
```

Comment:
This is probably fine, I'm just wondering why you went for doing a fresh dag vs just doing in place replacement with `subtitute_node_with_dag` for the dd sequence? With the in place approach you could probably just loop over `dag.named_nodes()` and work directly with only the delays. Did you benchmark to see which is faster?

Reply:
I haven't benchmarked this, agree it would be good to know. Probably as a follow up.